File: frontend/src/utils/betLabel.js
// Build a human-readable bet label from a trade or position.
//
// Uses the Kalshi ticker to extract team abbreviations (SEA, HOU, SD, COL, ...)
// instead of the full team name, so labels match the compact style of the
// scoreboard.
//
// Ticker patterns:
//   Moneyline:  KXMLBGAME-<date><away><home>-<HOME_ABBR>
//               e.g. KXMLBGAME-26APR112140HOUSEA-SEA  → home = SEA
//   Spread:     KXMLBSPREAD-<suffix>-<TEAM><N>
//               e.g. KXMLBSPREAD-26APR112040COLSD-COL4  → team = COL, N = 4
//   Over/Under: KXMLBTOTAL-<suffix>-<N>  (no team abbr needed)

export function inferMarketType(ticker) {
  if (!ticker) return null
  if (ticker.startsWith('KXMLBGAME')) return 'moneyline'
  if (ticker.startsWith('KXMLBTOTAL')) return 'over_under'
  if (ticker.startsWith('KXMLBSPREAD')) return 'spread'
  return null
}

function moneylineAbbr(ticker) {
  // Last segment after the final '-' is the home team abbr
  const parts = ticker?.split('-')
  if (!parts) return null
  const last = parts[parts.length - 1]
  return /^[A-Z]{2,4}$/.test(last) ? last : null
}

function spreadInfo(ticker) {
  // Last segment is like "COL4" or "SD5" — letters then integer N.
  // Kalshi convention: line = N - 0.5 (e.g. SEA4 → SEA -3.5)
  const parts = ticker?.split('-')
  if (!parts) return null
  const last = parts[parts.length - 1] || ''
  const match = last.match(/^([A-Z]{2,4})(\d+)$/)
  if (!match) return null
  return { abbr: match[1], line: parseInt(match[2], 10) - 0.5 }
}

function otherTeam(abbr, home, away) {
  if (!abbr) return null
  if (home && abbr === home) return away || null
  if (away && abbr === away) return home || null
  return null
}

export function betLabel(obj) {
  if (!obj) return null
  const ticker = obj.market_ticker || ''
  const side = obj.side
  const mt = obj.market_type || inferMarketType(ticker)
  const home = obj.home_abbr || ''
  const away = obj.away_abbr || ''

  if (!side) return null

  if (mt === 'over_under') {
    const linePart = ticker?.split('-').pop()
    const line = linePart?.match(/^\d+(?:\.\d+)?$/)?.[0]
    if (line) return side === 'YES' ? `Over ${line}` : `Under ${line}`
    return side === 'YES' ? 'Over' : 'Under'
  }

  if (mt === 'spread') {
    const info = spreadInfo(ticker)
    if (info) {
      if (side === 'YES') return `${info.abbr} -${info.line}`
      // NO on "SEA -4.5" = bet against SEA covering = "HOU +4.5"
      const flipped = otherTeam(info.abbr, home, away) || info.abbr
      return `${flipped} +${info.line}`
    }
    return `${side} Spread`
  }

  if (mt === 'moneyline') {
    const abbr = moneylineAbbr(ticker)
    if (abbr) {
      if (side === 'YES') return `${abbr} wins`
      // NO on "SEA wins" = bet SEA loses = "HOU wins"
      const flipped = otherTeam(abbr, home, away)
      return flipped ? `${flipped} wins` : `${abbr} loses`
    }
    return side === 'YES' ? 'Home wins' : 'Home loses'
  }

  return side
}

File: backend/app/bet_label.py
"""Human-readable bet label derivation.

Single source of truth so button grids, active position cards, and history
all render identical text for the same trade. Mirrors what the frontend
`betLabel` helper used to compute locally, now centralized server-side so
the frontend just reads `display_label`.
"""


def _spread_info(ticker: str) -> tuple[str, float] | None:
    """Parse `KXMLBSPREAD-...-{ABBR}{N}` → ({ABBR}, N-0.5). None if malformed."""
    if not ticker:
        return None
    last = ticker.rsplit("-", 1)[-1].upper()
    i = 0
    while i < len(last) and last[i].isalpha():
        i += 1
    abbr = last[:i]
    n_str = last[i:]
    if not abbr or not n_str:
        return None
    try:
        n = int(n_str)
    except ValueError:
        return None
    return abbr, n - 0.5


def _ou_line(ticker: str) -> str | None:
    """Parse `KXMLBTOTAL-...-{N}` → "{N}" (integer string)."""
    if not ticker:
        return None
    last = ticker.rsplit("-", 1)[-1]
    return last if last.isdigit() else None


def _ml_abbr(ticker: str) -> str | None:
    """Parse `KXMLBGAME-...-{HOME_ABBR}` → home abbr."""
    if not ticker:
        return None
    last = ticker.rsplit("-", 1)[-1].upper()
    return last if last.isalpha() and 2 <= len(last) <= 4 else None


def _infer_market_type(ticker: str) -> str | None:
    if not ticker:
        return None
    if ticker.startswith("KXMLBGAME"):
        return "moneyline"
    if ticker.startswith("KXMLBTOTAL"):
        return "over_under"
    if ticker.startswith("KXMLBSPREAD"):
        return "spread"
    return None


def compute_display_label(
    *,
    market_ticker: str,
    side: str,
    market_type: str | None = None,
    home_abbr: str = "",
    away_abbr: str = "",
) -> str:
    """Build the bet-meaning label. Flips team+sign on NO side so the
    user reads the side of the bet they actually placed.

    Spread: YES on `SEA -4.5` → "SEA -4.5"; NO → "HOU +4.5".
    Moneyline: YES on home ticker → "{HOME} wins"; NO → "{AWAY} wins".
    O/U: YES → "Over {N}"; NO → "Under {N}".
    """
    if not side:
        return ""
    mt = market_type or _infer_market_type(market_ticker) or ""
    home = (home_abbr or "").upper()
    away = (away_abbr or "").upper()

    if mt == "over_under":
        line = _ou_line(market_ticker)
        if line:
            return f"Over {line}" if side == "YES" else f"Under {line}"
        return "Over" if side == "YES" else "Under"

    if mt == "spread":
        info = _spread_info(market_ticker)
        if info:
            abbr, line = info
            line_str = f"{line:g}"
            if side == "YES":
                return f"{abbr} -{line_str}"
            other = home if abbr == away else (away if abbr == home else abbr)
            return f"{other} +{line_str}"

    if mt == "moneyline":
        abbr = _ml_abbr(market_ticker)
        if abbr:
            if side == "YES":
                return f"{abbr} wins"
            other = home if abbr == away else (away if abbr == home else None)
            return f"{other} wins" if other else f"{abbr} loses"
        return "Home wins" if side == "YES" else "Home loses"

    return side
